### ISSUE TITLE
Réduction stéréo : Volume en sortie correspondant à la loi de puissance

### ISSUE THREAD

**Issue by DavidPiazza** [GRIS-UdeM/SpatGRIS#382]
Version 3.2.2

Lorsqu'on est en réduction stéréo, changer le nombre de sources cause une modification au niveau du volume de sortie et ce, même si les sources sont inactives. Ce problème n'affecte que la réduction stéréo. Les modes binaural et multi-canal ne présentent pas de modification de volume.

Étapes pour la reproduction : 

- Créer un projet avec une source et y envoyer du signal;

- Placer la source à gauche ou à droite de l'espace et ajuster le gain de sortie à 0dB afin qu'un des deux VU mètres de sortie affiche le même niveau qu'en entrée;

<img width="1397" alt="une_source_hard_left_-6dBFS" src="https://user-images.githubusercontent.com/50223477/204675215-070eacf7-fe4e-4a70-a522-f91df3b7b072.png">

- Mettre le nombre de sources à 128;

- On remarque l'atténuation du volume d'environ -12dB : 

<img width="1397" alt="128sources_hard_left_-18dBFS" src="https://user-images.githubusercontent.com/50223477/204675436-fbc1a670-0788-4e06-951e-078d7064e104.png">

**Comment by Normandeau** [comment]
Je confirme le tout.

**Comment by NicolaGiannini** [comment]
Je confirme moi aussi.

**Comment by Normandeau** [comment]
Deux questions doivent être posées ici:

**1. Qui veut faire cela, ouvrir des entrées inutilisées en grand nombre en écoute stéréo?**
Essentiellement personne, surtout que c'est écrit noir sur blanc dans le manuel (bon, qui lit le manuel?) de pas dépenser inutilement des ressources inutilisées.

**2. N'est-ce pas exactement le comportement souhaité?**
Est-ce que la valeur de -12dB ne devrait sonner une grosse cloche dans votre esprit?
Je reprend le même dispositif que celui décrit plus haut mais avec 8 entrées au départ, et j'ajuste (arbitrairement et pour les besoins de la démonstration) la valeur de sortie sur un seul canal de la sortie stéréo à -20dB. Je double ensuite le nombre de sorties à chaque mesure. Voici les niveaux de sortie mesurées:
- 8: -20,0 dB
- 16: -20,8 dB
- 32: -22,4 dB
- 64: -25,6 dB
- 128: -32,0 dB

Donc effectivement -12dB de différence après avoir doublé le nombre de sorties 4 fois. Est-ce que cela n'éveille pas quelque chose en vous? Un souvenir du cours d'acoustique par exemple?
**La loi de puissance.**
Chaque fois que le nombre de sources est doublé, la puissance augmente de 3 dB. Donc, pour éviter de saturer la sortie stéréo avec cette augmentation de puissance, le mixeur stéréo compense de -3 dB pour chaque doublement du nombre de sources, conservant ainsi la sortie stéréo à une valeur stable.
Je vous laisse le soin d'en faire la preuve avec les 128 sources que cela suppose...

Donc c'est bien de tester mais il faut aller au bout de la démonstration...

**Comment by DavidPiazza** [comment]
> Deux questions doivent être posées ici:
> 
> **1. Qui veut faire cela, ouvrir des entrées inutilisées en grand nombre en écoute stéréo?** Essentiellement personne, surtout que c'est écrit noir sur blanc dans le manuel (bon, qui lit le manuel?) de pas dépenser inutilement des ressources inutilisées.

Si le début de ma pièce ne comprend qu'une piste qui envoie du signal à gauche, je m'attendrais à ce que le niveau que mon séquenceur envoie soit le même affiché dans SpatGRIS et ce, peu importe le nombre de sources dans le projet en entier. Les ressources sont belles et bien utilisées.

> Donc effectivement -12dB de différence après avoir doublé le nombre de sorties 4 fois. Est-ce que cela n'éveille pas quelque chose en vous? Un souvenir du cours d'acoustique par exemple? **La loi de puissance.** Chaque fois que le nombre de sources est doublé, la puissance augmente de 3 dB. Donc, pour éviter de saturer la sortie stéréo avec cette augmentation de puissance, le mixeur stéréo compense de -3 dB pour chaque doublement du nombre de sources, conservant ainsi la sortie stéréo à une valeur stable. Je vous laisse le soin d'en faire la preuve avec les 128 sources que cela suppose...
> 

Je connais très bien la loi de puissance et je comprends pourquoi on fait ça, mais je me demande s'il n'y aurait pas moyen d'activer où non cette compensation. Dans la situation où on est en train de mixer une pièce en stéréo, SpatGRIS agit comme un final output mixer et devrait logiquement avoir chacune de ses tranches au gain unitaire... Imaginez si dans votre DAW, l'ajout de pistes vides modifiaient invariablement le niveau de sorties des pistes qui contiennent des fichiers. Si je suis en train de réaliser le mix d'une pièce multi-canal en stéréo, les ajustements et les manipulations de la dynamique sont faits dans le séquenceur, pas dans SpatGRIS. Si ce dernier compense une augmentation de puissance alors que la puissance n'a pas réellement augmenté, je suggère qu'il faudrait peut-être spécifier clairement les caractéristiques de ce comportement dans le manuel (afin de pouvoir mixer en conséquence) et/ou potentiellement ajouter un interrupteur nous permettant d'enlever cette compensation?

**Comment by DavidPiazza** [comment]
> Deux questions doivent être posées ici:
> 
> **1. Qui veut faire cela, ouvrir des entrées inutilisées en grand nombre en écoute stéréo?** Essentiellement personne, surtout que c'est écrit noir sur blanc dans le manuel (bon, qui lit le manuel?) de pas dépenser inutilement des ressources inutilisées.

Si le début de ma pièce ne comprend qu'une piste qui envoie du signal à gauche, je m'attendrais à ce que le niveau que mon séquenceur envoie soit le même affiché dans SpatGRIS et ce, peu importe le nombre de sources dans le projet en entier. Les ressources sont belles et bien utilisées.

> Donc effectivement -12dB de différence après avoir doublé le nombre de sorties 4 fois. Est-ce que cela n'éveille pas quelque chose en vous? Un souvenir du cours d'acoustique par exemple? **La loi de puissance.** Chaque fois que le nombre de sources est doublé, la puissance augmente de 3 dB. Donc, pour éviter de saturer la sortie stéréo avec cette augmentation de puissance, le mixeur stéréo compense de -3 dB pour chaque doublement du nombre de sources, conservant ainsi la sortie stéréo à une valeur stable. Je vous laisse le soin d'en faire la preuve avec les 128 sources que cela suppose...
> 

Je connais très bien la loi de puissance et je comprends pourquoi on fait ça, mais je me demande s'il n'y aurait pas moyen d'activer où non cette compensation. Dans la situation où on est en train de mixer une pièce en stéréo, SpatGRIS agit comme un final output mixer et devrait logiquement avoir chacune de ses tranches au gain unitaire... Imaginez si dans votre DAW, l'ajout de pistes vides modifiaient invariablement le niveau de sorties des pistes qui contiennent des fichiers. Si je suis en train de réaliser le mix d'une pièce multi-canal en stéréo, les ajustements et les manipulations de la dynamique sont faits dans le séquenceur, pas dans SpatGRIS. Si ce dernier compense une augmentation de puissance alors que la puissance n'a pas réellement augmenté, je suggère qu'il faudrait peut-être spécifier clairement les caractéristiques de ce comportement dans le manuel (afin de pouvoir mixer en conséquence) et/ou potentiellement ajouter un interrupteur nous permettant d'enlever cette compensation?

**Comment by Normandeau** [comment]
Oui, bonne question. D'autant plus que la loi de puissance ici invoquée n'est appliquée que dans le cas de pistes vides. Dès qu'il y a du signal, le niveau de sortie augmente comme prévu. Testé seulement avec 8 entrées, bruit rose et sinus 1k.
L'algorithme Stereo devait être testé et revu.

**Comment by NicolaGiannini** [comment]
V326

Le comportement me semble le même de celui décrit au début de l’issue. Cela dit, d’une certaine manière, je pense qu'il est probablement utile que SpatGRIS diminue le volume global de sortie en fonction du nombre de sources. Comme Robert l'a fait remarquer, c'est probablement pour ne pas saturer rapidement le master lorsque toutes les sources sont envoyées à seulement deux haut-parleurs. Si on considère la réduction stéréo comme quelque chose que l'on utilise uniquement lorsque le projet est terminé pour créer une réduction stéréo, je pense que le volume global peut alors être ajusté en conséquence dans ce contexte. Car si je comprends bien, SpatGRIS diminue le volume global, mais le mixage des sources reste le même, n'est-ce pas ?

**Comment by DavidPiazza** [comment]
>  Car si je comprends bien, SpatGRIS diminue le volume global, mais le mixage des sources reste le même, n'est-ce pas ?

Oui, SpatGRIS diminue le volume global en fonction du nombre de sources dans le projet, indépendamment du fait qu'elles soient actives ou non. Le niveau absolu des sources est abaissé, ce qui fait qu'on a nécessairement une modification proportionnelle de la dynamique de la pièce. Je trouverais ça bizarre que mon séquenceur baisse le niveau de toutes les pistes quand j'en ajoute des vides...

J'imagine que la solution simple est de compenser cette modification avec le potentiomètre du gain de sortie.

**Comment by Normandeau** [comment]
V401
Bonne discussion. On pourra y revenir.
La situation n'est pas changée dans la plus récente version. Globalement, SpatGRIS est «prudent» dans le sens où si des pistes sont ajoutées, qu'elles soient active ou pas, le niveau général de la sortie stéréo est diminuée. Ça évite d'endommager des hp!!!
